labview-interop/src/errors.rs

```diff
@@ -1,89 +1,650 @@
+        if self == Self::SUCCESS {
+            Ok(success_value)
         } else {
+            match MgError::try_from(self) {
+                Ok(mg_err) => Err(mg_err.into()),
+                Err(inter_err) => Err(inter_err),
+            }
+        }
+    }
+
+    /// this will convert the LVStatusCode to the generic LVError with no checks of validity
+    pub fn to_generic_result<T>(self, success_value: T) -> core::result::Result<T, LVError> {
+        if self == Self::SUCCESS {
             Ok(success_value)
+        } else {
+            Err(LVError { code: self })
         }
     }
+}
+
+impl From<LVError> for LVStatusCode {
+    fn from(err: LVError) -> LVStatusCode {
+        err.code
+    }
+}
 
-    fn get_description(&self) -> &'static str {
-        match self.0 {
-            0 => "No Error",
-            2 => "Memory Full",
-            _ => "No Description for Code",
+impl<T> From<core::result::Result<T, LVError>> for LVStatusCode {
+    fn from(value: core::result::Result<T, LVError>) -> Self {
+        match value {
+            Ok(_) => LVStatusCode::SUCCESS,
+            Err(err) => err.into(),
         }
     }
 }
 
-impl Display for MgErr {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}: {}", self.0, self.get_description())
+// From<i32> vice versa implemented, but not Deref (do not want to inherit other math operations)
+impl From<i32> for LVStatusCode {
+    fn from(value: i32) -> LVStatusCode {
+        LVStatusCode(value)
     }
 }
 
-impl Error for MgErr {
-    fn source(&self) -> Option<&(dyn Error + 'static)> {
-        None
+impl From<LVStatusCode> for i32 {
+    fn from(code: LVStatusCode) -> i32 {
+        code.0
     }
 }
 
-#[derive(Error, Debug)]
-pub enum LVInteropError {
-    #[error("Internal LabVIEW Error: {0}")]
-    LabviewError(#[from] MgErr),
-    #[error("Invalid handle when valid handle is required")]
-    InvalidHandle,
+impl Display for LVStatusCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// the conversion from LVInteropError back to LVStatusCode is important
+/// to return the status in extern "C" functions back to LV
+impl<T> From<Result<T>> for LVStatusCode {
+    fn from(value: Result<T>) -> Self {
+        match value {
+            Ok(_) => LVStatusCode::SUCCESS,
+            Err(err) => err.into(),
+        }
+    }
+}
+impl From<LVInteropError> for LVStatusCode {
+    fn from(value: LVInteropError) -> Self {
+        let num = match value {
+            LVInteropError::LabviewMgError(e) => e as i32,
+            LVInteropError::InternalError(e) => e as i32, // TODO
+        };
+        num.into()
+    }
+}
+
+/// LVError is a generic Labview Error
+/// that can retrieve a error description if the link feature is enabled
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Error)]
+pub struct LVError {
```

Comment:
I'm not very convinced that this additional level adds anything over the StatusCode. I would probably remove this